### ISSUE TITLE
Access to the data inside functions [enhancement/question]

### ISSUE THREAD

**Issue by polderudo** [maja42/goval#10]
Is there any way to get access to the passed in variables/data inside the functions? At least readonly access would be great.

**Comment by maja42** [comment]
I'm not sure I correctly understand your question. You can access variables if they are passed as an argument (readonly).
If you want (read+write) access to all variables, even if they are not passed directly, you can use closures as functions.
Does this answer your question?

**Comment by polderudo** [comment]
We are evaluating like this:
`fret, err := eval.Evaluate(r.Rule, data, functions)`

rule is the expression to be evaluated, data the data that can be used for the evaluation and functions the functions we can use in the rule.

functions are like:
```
f["HideIf"] = func(args ...interface{}) (interface{}, error) {
		//only one argument if hide or show
		if len(args) != 1 {
			return nil, fmt.Errorf("wrong number of args for Hide")
		}

		hide := args[0].(bool)

		hideCnt++
		return FuncReturn{
			Meta: &FuncMeta{
				Hidden: &hide,
			},
		}, nil
	}
```
My problem is, that i can't access dat inside HideIf.

The data of course is available inside the evaluation, but not in the functions.

BTW: Very nice library :) It's realy fast and we are able to build nice, editable business logic with it, with hundrets of small "rules" that evaluate in milliseconds.

**Comment by maja42** [comment]
Thanks, that's nice to hear :)

The library itself doesn't pass you all the variables. But it doesn't need to. You can just use closures.
As long as the functions are defined in the same scope as the variables, they can access it.

Here's some pseudo-example-code. You can do it like this:

```go
func Evaluate() (interface{}, error) {
    eval := goval.NewEvaluator()
    variables := map[string]interface{}{
        "var": 42,
    }

    functions := map[string]goval.ExpressionFunction {
        "fun": func(args ...interface{}) (interface{}, error) {
            // access "variables" here
            return variables["var"], nil 
        },
    }

    return eval.Evaluate(`fun()`, variables, functions)  // Returns <42, nil>
}
```

Alternatively, a common go-idiom is also something like this:

```go
func GetFunctions(variables map[string]interface{}) map[string]goval.ExpressionFunction {
    return map[string]goval.ExpressionFunction {
        "fun": func(args ...interface{}) (interface{}, error) {
            // access "variables" here
            return variables["var"], nil 
        },
    }
}

func Evaluate() (interface{}, error) {
    eval := goval.NewEvaluator()
    variables := map[string]interface{}{
        "var": 42,
    }
    functions := map[string]goval.ExpressionFunction {
        // some generic functions that don't need access to variables
    }
    functions = append(functions, GetFunctions(variables))

    return eval.Evaluate(`fun()`, variables, functions)  // Returns <42, nil>
}
```

And, last but not least, you don't even need access to the variables directly. You can encapsulate it via an interface:

```go
type VarAccess interface {
    Get(varName string) (interface{}, bool)
}

func GetFunctions(variables VarAccess) map[string]goval.ExpressionFunction {
    return map[string]goval.ExpressionFunction {
        "fun": func(args ...interface{}) (interface{}, error) {
            val, ok := variables.Get("var")
            _ = ok
            return val, nil 
        },
    }
}

type Variables map[string]interface{}

func (v *Variables) Get(varName string) (interface{}, bool) {
    return v[varName]
}
```

TL;DR: Go already offers you all the tools you need via its closures.

**Comment by polderudo** [comment]
Man ... you are right. Closures do solve this. I forgot about this and made the functions completly stateless, but i don't need to of course :) Closing the issue. Thanks for opening the eyes.
